src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -62,6 +66,171 @@ public static Document Current
+                {
+                    var compundStructure = hostObj.GetCompoundStructure();
+                    if (compundStructure == null)
+                        continue;
+
+                    int layerCount = compundStructure.LayerCount;
+                    for (int j = 0; j < layerCount; j++)
+                    {
+                        ElementId elementId = compundStructure.GetMaterialId(j);
+                        if (materials.Contains(elementId))
+                        {
+                            nonPurgedElements.Add(elementId);
+                            materials.Remove(elementId);
+                        }
+                        
+                    }
+                    continue;
+                }
+
+                List<ElementId> elementMaterialIds = elementTypes[i].GetMaterialIds(false).ToList();
+                if (elementMaterialIds.Any())
+                {
+                    materials.RemoveAll(purgedId => elementMaterialIds.Exists(elemId => purgedId == elemId));
+                    nonPurgedElements.AddRange(elementMaterialIds.Where(elmeMatId => !nonPurgedElements.Any(nonPurgedElemeMatId => nonPurgedElemeMatId == elmeMatId)));
+                }
+
+                if (materials == null)
+                    break;
+            }
+            if (materials.Count > 0)
+                this.InternalDocument.Delete(materials);
+            List<ElementId> purgedMaterials = new List<ElementId>();
+            purgedMaterials.AddRange(PurgeMaterialAssets(nonPurgedElements));
+            purgedMaterials.AddRange(materials);
+
+            return purgedMaterials;
+        }
+
+        private List<Autodesk.Revit.DB.ElementId> PurgeMaterialAssets(List<Autodesk.Revit.DB.ElementId> elementIds)
+        {
+            List<ElementId> appearanceAssets = new FilteredElementCollector(this.InternalDocument).OfClass(typeof(AppearanceAssetElement))
+                                                                                      .ToElementIds().ToList();
+
+            List<ElementId> propertySet = new FilteredElementCollector(this.InternalDocument).OfClass(typeof(PropertySetElement))
+                                                                                                      .ToElementIds().ToList();
+
+            for (int i = 0; i < elementIds.Count(); i++)
+            {
+                Autodesk.Revit.DB.Material material = this.InternalDocument.GetElement(elementIds[i]) as Autodesk.Revit.DB.Material;
+                propertySet.Remove(material.ThermalAssetId);
+                propertySet.Remove(material.StructuralAssetId);
+                appearanceAssets.Remove(material.AppearanceAssetId);
+            }
+
+            if (appearanceAssets.Count > 0)
+                this.InternalDocument.Delete(appearanceAssets);
+            if (propertySet.Count > 0)
+                this.InternalDocument.Delete(propertySet);
+
+            List<ElementId> purgedElementIds = new List<ElementId>();
```

Comment:
you can pass in a list to the constructor to make the list & add things to it at the same time 😉